### ISSUE TITLE
Generalize is_categorical_dtype to support alternative backends

### ISSUE THREAD

**Issue by beckernick** [dask/dask#7054]
Currently, Dask relies on `pd.api.types.is_categorical_dtype` to verify whether a column is categorical dtype or not. This is used in various places across the codebase. Columns backed by non-pandas backends may not be able to pass this check (cuDF cannot), which can cause errors using at least some functionality (`get_dummies`).

```python
import pandas as pd
s = pd.Series(["a","b","c"]).astype("category")
pd.api.types.is_categorical_dtype(s) # True

import cudf
s = cudf.Series(["a","b","c"]).astype("category")
cudf.utils.dtypes.is_categorical_dtype(s) # True
pd.api.types.is_categorical_dtype(s) # False
```
One example usage is in the `get_dummies` API.

https://github.com/dask/dask/blob/51add4e5c07fd0deb1b8cc179c7d66db9f8cc063/dask/dataframe/reshape.py#L132-L147

This works as expected for the pandas backed dataframe, but fails with the following for cuDF backed dataframes.

```python
import pandas as pd
import dask.dataframe as dd
import cudf
​
df = pd.DataFrame(
    {"A":["a","b","b"],
     "B":[1,2,3]
    })
ddf = dd.from_pandas(df, 2)
ddf = ddf.map_partitions(cudf.from_pandas)
ddf = ddf.categorize(columns=["B"])
dd.get_dummies(ddf, columns=['B'])
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-9-ec7ee286026f> in <module>
     10 ddf = ddf.map_partitions(cudf.from_pandas)
     11 ddf = ddf.categorize(columns=["B"])
---> 12 dd.get_dummies(ddf, columns=['B'])

/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-automated-tests/lib/python3.7/site-packages/dask/dataframe/reshape.py in get_dummies(data, prefix, prefix_sep, dummy_na, columns, sparse, drop_first, dtype, **kwargs)
    142         else:
    143             if not all(is_categorical_dtype(data[c]) for c in columns):
--> 144                 raise NotImplementedError(not_cat_msg)
    145 
    146         if not all(has_known_categories(data[c]) for c in columns):

NotImplementedError: `get_dummies` with non-categorical dtypes is not supported. Please use `df.categorize()` beforehand to convert to categorical dtype.
```

This is related to https://github.com/rapidsai/cudf/issues/7111 and may be relevant for https://github.com/rapidsai/cudf/pull/6160

Environment
```
conda list | grep "rapids\|dask"
cudf                      0.18.0a210110   cuda_10.2_py37_g04aa30cf11_157    rapidsai-nightly
cuml                      0.18.0a210110   cuda10.2_py37_gc021decd9_70    rapidsai-nightly
dask                      2020.12.0          pyhd8ed1ab_0    conda-forge
dask-core                 2020.12.0          pyhd8ed1ab_0    conda-forge
dask-cuda                 0.18.0a201211           py37_39    rapidsai-nightly
dask-cudf                 0.18.0a210110   py37_g04aa30cf11_157    rapidsai-nightly
faiss-proc                1.0.0                      cuda    rapidsai-nightly
libcudf                   0.18.0a210110   cuda10.2_g04aa30cf11_157    rapidsai-nightly
libcuml                   0.18.0a210110   cuda10.2_gc021decd9_70    rapidsai-nightly
libcumlprims              0.18.0a201203   cuda10.2_gff080f3_0    rapidsai-nightly
librmm                    0.18.0a210110   cuda10.2_g94b083a_20    rapidsai-nightly
rmm                       0.18.0a210110   cuda_10.2_py37_g94b083a_20    rapidsai-nightly
ucx                       1.8.1+g6b29558       cuda10.2_0    rapidsai-nightly
ucx-proc                  1.0.0                       gpu    rapidsai-nightly
ucx-py                    0.18.0a210110   py37_g6b29558_10    rapidsai-nightly
```

**Comment by TomAugspurger** [comment]
Can pandas or cudf be updated to make this work? https://github.com/pandas-dev/pandas/blob/6e707f2f68a4ed7fd990438bc489374641374cff/pandas/core/dtypes/common.py#L537

It looks like if cudf's dtypes inherit from `pandas.api.types.ExtensionDtype` then this would work. Is that an option?

**Comment by beckernick** [comment]
Good thought. Possibly. Digging a bit into rapidsai/cudf#6160, this may actually be covered by this PR already. Will circle back to this issue and provide an update once the status is more clear.

**Comment by brandon-b-miller** [comment]
I think the check passes for the actual `cudf.CategoricalDtype` objects, it's just that it doesn't for a `cudf.Series` of that dtype. 

```
>>> dt = cudf.CategoricalDtype(categories=[1,2,3])
>>> pd.api.types.is_categorical_dtype(dt)
True
```

I am a little surprised since I would expect it to hit [this line](https://github.com/pandas-dev/pandas/blob/6e707f2f68a4ed7fd990438bc489374641374cff/pandas/core/dtypes/base.py#L278) in pandas and return true even for the cuDF series, so maybe a pandas bug?

All that said, I am glad that you proposed this, as this more or less came up when we were experimenting with `cudf.dtype`. We realized that something like this might have to happen for a section of the numpy and possibly pandas APIs for dask to keep working with cuDF. 

Basically going from:

```
pd.api.types.is_categorical_dtype(...)
np.dtype(...)
```

to:

```
import cudf as module
module.api.types.is_categorical_dtype(...)
module.dtype(...)
```

To that end I'm curious what the implications would be on the dask side and to hear peoples opinions on this sort of generalization.

**Comment by brandon-b-miller** [comment]
Hi all,
Having looked into this a bit, it seems to me the problem motivating this issue could be fixed in any of cuDF, Pandas, or Dask. To sort out exactly what to do, I collected the details of the problem in https://github.com/rapidsai/cudf/issues/7111 and am hoping to generate some consensus there on how to proceed.

**Comment by jsignell** [comment]
Closed by #7469
